File: internal/consolidation/autoprune.go
package consolidation

import (
	"context"
	"fmt"
	"time"

	"github.com/cortex-ai/cortex-ai/internal/config"
	"github.com/cortex-ai/cortex-ai/internal/memory"
	"github.com/cortex-ai/cortex-ai/internal/search"
	"github.com/cortex-ai/cortex-ai/internal/storage"
)

// AutopruneService handles automatic memory cleanup and optimization
type AutopruneService struct {
	storage  storage.Storage
	embedder memory.Embedder
	config   *config.AutopruneConfig
}

// NewAutopruneService creates a new autoprune service
func NewAutopruneService(store storage.Storage, embedder memory.Embedder, cfg *config.AutopruneConfig) *AutopruneService {
	return &AutopruneService{
		storage:  store,
		embedder: embedder,
		config:   cfg,
	}
}

// AutopruneOptions configures which cleanup operations to perform
type AutopruneOptions struct {
	Duplicates      bool          // remove duplicate memories
	ArchiveEpisodic bool          // archive old episodic memories
	MergeSemantic   bool          // merge similar semantic memories
	DryRun          bool          // simulate without making changes
	OlderThan       time.Duration // age threshold for archiving
}

// AutopruneResult contains the results of an autoprune operation
type AutopruneResult struct {
	DuplicatesRemoved int      `json:"duplicates_removed"`
	EpisodicArchived  int      `json:"episodic_archived"`
	SemanticMerged    int      `json:"semantic_merged"`
	Details           []string `json:"details,omitempty"`
}

// Run executes the autoprune operation based on options
func (s *AutopruneService) Run(ctx context.Context, opts AutopruneOptions) (*AutopruneResult, error) {
	result := &AutopruneResult{
		Details: make([]string, 0),
	}

	// If no specific option is set, run all
	runAll := !opts.Duplicates && !opts.ArchiveEpisodic && !opts.MergeSemantic

	if opts.Duplicates || runAll {
		count, details, err := s.removeDuplicates(ctx, opts.DryRun)
		if err != nil {
			return nil, fmt.Errorf("failed to remove duplicates: %w", err)
		}
		result.DuplicatesRemoved = count
		result.Details = append(result.Details, details...)
	}

	if opts.ArchiveEpisodic || runAll {
		olderThan := opts.OlderThan
		if olderThan == 0 {
			olderThan = time.Duration(s.config.EpisodicRetentionDays) * 24 * time.Hour
		}
		count, details, err := s.archiveEpisodic(ctx, olderThan, opts.DryRun)
		if err != nil {
			return nil, fmt.Errorf("failed to archive episodic: %w", err)
		}
		result.EpisodicArchived = count
		result.Details = append(result.Details, details...)
	}

	if opts.MergeSemantic || runAll {
		count, details, err := s.mergeSemantic(ctx, opts.DryRun)
		if err != nil {
			return nil, fmt.Errorf("failed to merge semantic: %w", err)
		}
		result.SemanticMerged = count
		result.Details = append(result.Details, details...)
	}

	return result, nil
}

// removeDuplicates finds and removes duplicate memories across all levels
func (s *AutopruneService) removeDuplicates(ctx context.Context, dryRun bool) (int, []string, error) {
	var details []string
	removed := 0

	for _, level := range memory.ValidMemoryLevels {
		memories, err := s.storage.List(ctx, memory.ListOptions{FilterLevels: []memory.MemoryLevel{level}})
		if err != nil {
			continue
		}

		seen := make(map[string]bool)
		for i, m1 := range memories {
			if seen[m1.ID] {
				continue
			}

			for j := i + 1; j < len(memories); j++ {
				m2 := memories[j]
				if seen[m2.ID] {
					continue
				}

				similarity := search.CosineSimilarity(m1.Embedding, m2.Embedding)
				if similarity >= s.config.DuplicatesThreshold {
					if dryRun {
						details = append(details, fmt.Sprintf("[dry-run] would remove duplicate: %s (similar to %s, score: %.3f)", m2.ID[:8], m1.ID[:8], similarity))
					} else {
						if err := s.storage.Delete(ctx, m2.ID); err == nil {
							details = append(details, fmt.Sprintf("removed duplicate: %s (similar to %s)", m2.ID[:8], m1.ID[:8]))
							removed++
						}
					}
					seen[m2.ID] = true
				}
			}
		}
	}

	return removed, details, nil
}

// archiveEpisodic archives episodic memories older than the threshold
func (s *AutopruneService) archiveEpisodic(ctx context.Context, olderThan time.Duration, dryRun bool) (int, []string, error) {
	var details []string
	before := time.Now().Add(-olderThan)

	memories, err := s.storage.List(ctx, memory.ListOptions{FilterLevels: []memory.MemoryLevel{memory.MemoryLevelEpisodic}})
	if err != nil {
		return 0, nil, err
	}

	count := 0
	for _, m := range memories {
		if m.CreatedAt.Before(before) {
			if dryRun {
				details = append(details, fmt.Sprintf("[dry-run] would archive: %s (created: %s)", m.ID[:8], m.CreatedAt.Format("2006-01-02")))
				count++
				continue
			}
			if err := s.storage.Delete(ctx, m.ID); err == nil {
				count++
			}
		}
	}

	if count > 0 && !dryRun {
		details = append(details, fmt.Sprintf("archived %d episodic memories older than %v", count, olderThan))
	}

	return count, details, nil
}

// mergeSemantic merges similar semantic memories
func (s *AutopruneService) mergeSemantic(ctx context.Context, dryRun bool) (int, []string, error) {
	var details []string
	merged := 0

	memories, err := s.storage.List(ctx, memory.ListOptions{FilterLevels: []memory.MemoryLevel{memory.MemoryLevelSemantic}})
	if err != nil {
		return 0, nil, err
	}

	mergedIDs := make(map[string]bool)

	for i, m1 := range memories {
		if mergedIDs[m1.ID] {
			continue
		}

		var toMerge []*memory.Memory
		for j := i + 1; j < len(memories); j++ {
			m2 := memories[j]
			if mergedIDs[m2.ID] {
				continue
			}

			similarity := search.CosineSimilarity(m1.Embedding, m2.Embedding)
			if similarity >= s.config.SemanticMergeThreshold {
				toMerge = append(toMerge, m2)
				mergedIDs[m2.ID] = true
			}
		}

		if len(toMerge) > 0 {
			if dryRun {
				ids := make([]string, len(toMerge))
				for i, m := range toMerge {
					ids[i] = m.ID[:8]
				}
				details = append(details, fmt.Sprintf("[dry-run] would merge %d memories into %s", len(toMerge), m1.ID[:8]))
				continue
			}

			mergedContent := m1.Content
			sourceIDs := make([]string, 0, len(toMerge))
			for _, m := range toMerge {
				mergedContent += "\n\n---\n\n" + m.Content
				sourceIDs = append(sourceIDs, m.ID)
			}

			newEmbedding, err := s.embedder.Embed(ctx, mergedContent)
			if err != nil {
				continue
			}

			m1.Content = mergedContent
			m1.Embedding = newEmbedding
			m1.UpdatedAt = time.Now()
			m1.MergedFrom = append(m1.MergedFrom, sourceIDs...)
			m1.Tags = mergeTags(m1.Tags, collectTags(toMerge))
			m1.Context.Tags = mergeTags(m1.Context.Tags, collectContextTags(toMerge))

			if err := s.storage.Update(ctx, m1); err != nil {
				continue
			}
			for _, m := range toMerge {
				_ = s.storage.Delete(ctx, m.ID)
			}

			details = append(details, fmt.Sprintf("merged %d memories into %s", len(toMerge), m1.ID[:8]))
			merged += len(toMerge)
		}
	}

	return merged, details, nil
}

func collectTags(memories []*memory.Memory) []string {
	var tags []string
	for _, m := range memories {
		tags = append(tags, m.Tags...)
	}
	return tags
}

func collectContextTags(memories []*memory.Memory) []string {
	var tags []string
	for _, m := range memories {
		tags = append(tags, m.Context.Tags...)
	}
	return tags
}

File: src/cortex/consolidation/autoprune.py
"""Autoprune service - automated memory maintenance and optimization."""

from __future__ import annotations

import contextlib
from datetime import UTC, datetime, timedelta

import numpy as np

from cortex.models.memory import Memory, MemoryLevel
from cortex.models.results import AutopruneResult, AutopruneStats
from cortex.search.cosine import cosine_similarity, normalize
from cortex.storage.chroma import ChromaStorage

_MERGE_SEPARATOR = "\n\n---\n\n"


class AutopruneService:
    """Runs three automated cleanup operations:

    1. Remove duplicates - pairwise similarity >= duplicates_threshold (0.92)
    2. Archive episodic - delete episodic memories older than retention_days (90)
    3. Merge semantic - merge very similar semantic memories (>= semantic_merge_threshold 0.88)
    """

    def __init__(
        self,
        storage: ChromaStorage,
        duplicates_threshold: float = 0.92,
        episodic_retention_days: int = 90,
        semantic_merge_threshold: float = 0.88,
    ) -> None:
        self._storage = storage
        self._dup_threshold = duplicates_threshold
        self._retention_days = episodic_retention_days
        self._semantic_threshold = semantic_merge_threshold

    def run(self, dry_run: bool = False) -> AutopruneResult:
        """Execute all three autoprune operations in sequence."""
        result = AutopruneResult(dry_run=dry_run)
        result.duplicates = self._remove_duplicates(dry_run)
        result.episodic = self._archive_episodic(dry_run)
        result.semantic = self._merge_semantic(dry_run)
        return result

    # ------------------------------------------------------------------
    # Operation 1: Remove duplicates
    # ------------------------------------------------------------------

    def _remove_duplicates(self, dry_run: bool) -> AutopruneStats:
        """Find and remove near-duplicate memories across all levels."""
        stats = AutopruneStats()

        for level in [MemoryLevel.episodic, MemoryLevel.semantic]:
            items = self._storage.get_all_with_embeddings(level)
            seen_ids: set[str] = set()
            deleted_ids: set[str] = set()

            for i, (mem_a, vec_a) in enumerate(items):
                if mem_a.id in deleted_ids or not vec_a:
                    continue

                for mem_b, vec_b in items[i + 1 :]:
                    if mem_b.id in deleted_ids or not vec_b:
                        continue
                    if mem_b.id in seen_ids:
                        continue

                    score = cosine_similarity(vec_a, vec_b)
                    if score >= self._dup_threshold:
                        # Delete the one updated less recently
                        to_delete = mem_b if mem_a.updated_at >= mem_b.updated_at else mem_a
                        if not dry_run:
                            with contextlib.suppress(KeyError):
                                self._storage.delete(to_delete.id)
                        deleted_ids.add(to_delete.id)
                        stats.removed += 1
                        stats.details.append(
                            f"Removed duplicate {to_delete.id[:8]} (score={score:.3f})"
                        )

                seen_ids.add(mem_a.id)

        return stats

    # ------------------------------------------------------------------
    # Operation 2: Archive episodic memories
    # ------------------------------------------------------------------

    def _archive_episodic(self, dry_run: bool) -> AutopruneStats:
        """Delete episodic memories older than the retention threshold."""
        stats = AutopruneStats()
        cutoff = datetime.now(UTC) - timedelta(days=self._retention_days)

        items = self._storage.get_all_with_embeddings(MemoryLevel.episodic)
        for mem, _ in items:
            if mem.created_at < cutoff and not mem.obsolete:
                if not dry_run:
                    with contextlib.suppress(KeyError):
                        self._storage.delete(mem.id)
                stats.archived += 1
                stats.details.append(
                    f"Archived episodic {mem.id[:8]} (age={mem.created_at.date()})"
                )

        return stats

    # ------------------------------------------------------------------
    # Operation 3: Merge similar semantic memories
    # ------------------------------------------------------------------

    def _merge_semantic(self, dry_run: bool) -> AutopruneStats:
        """Merge pairs of very similar semantic memories."""
        stats = AutopruneStats()
        items = self._storage.get_all_with_embeddings(MemoryLevel.semantic)
        merged_ids: set[str] = set()

        for i, (mem_a, vec_a) in enumerate(items):
            if mem_a.id in merged_ids or not vec_a:
                continue

            for mem_b, vec_b in items[i + 1 :]:
                if mem_b.id in merged_ids or not vec_b:
                    continue

                score = cosine_similarity(vec_a, vec_b)
                if score >= self._semantic_threshold:
                    if not dry_run:
                        self._do_merge_semantic(mem_a, vec_a, mem_b, vec_b)
                    merged_ids.add(mem_b.id)
                    stats.merged += 1
                    stats.details.append(
                        f"Merged semantic {mem_b.id[:8]} into {mem_a.id[:8]} (score={score:.3f})"
                    )

        return stats

    def _do_merge_semantic(
        self,
        target: Memory,
        target_vec: list[float],
        source: Memory,
        source_vec: list[float],
    ) -> None:
        """Merge source into target, deleting source."""
        target.content = target.content + _MERGE_SEPARATOR + source.content
        target.tags = _dedupe_list(target.tags + source.tags)
        target.merged_from = _dedupe_list([*target.merged_from, source.id, *source.merged_from])
        target.touch()

        # Average the two embeddings
        averaged = np.mean([target_vec, source_vec], axis=0).tolist()
        target.embedding = normalize(averaged)

        self._storage.update(target)
        with contextlib.suppress(KeyError):
            self._storage.delete(source.id)


def _dedupe_list(items: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for item in items:
        if item not in seen:
            seen.add(item)
            result.append(item)
    return result
